### PR TITLE
Improve unit testing

### PR DESCRIPTION
**Pull request by DMHP**
### Proposed changes in this pull request
Writing unit test cases

### When should this PR be merged
Immediate

### Follow up actions
None

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.user.endpoint/pom.xml [hunk 1]
```diff
@@ -117,7 +117,7 @@
                                         <limit implementation="org.jacoco.report.check.Limit">
                                             <counter>COMPLEXITY</counter>
                                             <value>COVEREDRATIO</value>
-                                            <minimum>0.5</minimum>
+                                            <minimum>0.6</minimum>
```
Comment (kasunbg) [inline components/org.wso2.carbon.identity.user.endpoint/pom.xml]: Nice one Hasanthi!! :+1:

File: components/org.wso2.carbon.identity.user.endpoint/src/test/java/org/wso2/carbon/identity/user/endpoint/impl/MeApiServiceImplTest.java [hunk 2]
```diff
@@ -63,20 +64,44 @@
...
 
     @Test
-    public void testResendCodePost() throws Exception {
-
-        mockStatic(IdentityUtil.class);
-        mockStatic(Utils.class);
-        PowerMockito.when(IdentityUtil.class, (String) IdentityUtil.threadLocalProperties.get().
-                get(Constants.TENANT_NAME_FROM_CONTEXT)).thenReturn("TestTenant");
-        when(Utils.getUserSelfRegistrationManager()).thenReturn(userSelfRegistrationManager);
+    public void testMePost() throws Exception {
```
Comment (kasunbg) [inline components/org.wso2.carbon.identity.user.endpoint/src/test/java/org/wso2/carbon/identity/user/endpoint/impl/MeApiServiceImplTest.java]: missing method comments.

File: components/org.wso2.carbon.identity.user.endpoint/src/test/java/org/wso2/carbon/identity/user/endpoint/impl/ValidateCodeApiServiceImplTest.java [hunk 3]
```diff
@@ -43,14 +47,41 @@
...
 
+    @Mock
+    private Throwable throwable;
+
     @InjectMocks
     private ValidateCodeApiServiceImpl validateCodeApiServiceImpl;
 
     @Test
-    public void testValidateCodePost() {
+    public void testValidateCodePost() throws Exception{
```
Comment (kasunbg) [inline components/org.wso2.carbon.identity.user.endpoint/src/test/java/org/wso2/carbon/identity/user/endpoint/impl/ValidateCodeApiServiceImplTest.java]: code formatting is needed. Why are we throwing generic exceptions anyway?

File: components/org.wso2.carbon.identity.user.endpoint/src/test/java/org/wso2/carbon/identity/user/endpoint/impl/ValidateCodeApiServiceImplTest.java [hunk 4]
```diff
@@ -43,14 +47,41 @@
...
+
+        mockClasses();
+        assertEquals(validateCodeApiServiceImpl.validateCodePost(createCodeValidationRequestDTO()).getStatus(), 202);
+        assertEquals(validateCodeApiServiceImpl.validateCodePost(null).getStatus(), 202);
+    }
```
Comment (kasunbg) [inline components/org.wso2.carbon.identity.user.endpoint/src/test/java/org/wso2/carbon/identity/user/endpoint/impl/ValidateCodeApiServiceImplTest.java]: add a new line.

File: components/org.wso2.carbon.identity.user.endpoint/src/test/java/org/wso2/carbon/identity/user/endpoint/impl/ValidateCodeApiServiceImplTest.java [hunk 5]
```diff
@@ -43,14 +47,41 @@
     @Mock
     private UserSelfRegistrationManager userSelfRegistrationManager;
 
+    @Mock
+    private Throwable throwable;
```
Comment (kasunbg) [inline components/org.wso2.carbon.identity.user.endpoint/src/test/java/org/wso2/carbon/identity/user/endpoint/impl/ValidateCodeApiServiceImplTest.java]: Why did we have to mock throwable? I do not see this anywhere as well. :)

File: components/org.wso2.carbon.identity.user.endpoint/src/test/java/org/wso2/carbon/identity/user/endpoint/impl/ValidateCodeApiServiceImplTest.java [hunk 6]
```diff
@@ -43,14 +47,41 @@
...
         mockStatic(Utils.class);
         when(Utils.getUserSelfRegistrationManager()).thenReturn(userSelfRegistrationManager);
+    }
+
+    @Test
+    public void testIdentityRecoveryExceptioninResendCodePost() throws Exception {
+
+        mockClasses();
+        doThrow(new IdentityRecoveryException("Recovery Exception")).when(userSelfRegistrationManager).
+                confirmUserSelfRegistration(anyString());
```
Comment (kasunbg) [inline components/org.wso2.carbon.identity.user.endpoint/src/test/java/org/wso2/carbon/identity/user/endpoint/impl/ValidateCodeApiServiceImplTest.java]: Should you not have expectedExceptions definition when we do doThrow?

@Test(expectedExceptions = IllegalStateException.class)

**Comment by kasunbg** [comment]
Brnch seem to have conflicts.. Also, can u fill the PR template?

File: components/org.wso2.carbon.identity.user.endpoint/src/test/java/org/wso2/carbon/identity/user/endpoint/impl/MeApiServiceImplTest.java (hunk 2 above)
Comment (DMHP) [reply components/org.wso2.carbon.identity.user.endpoint/src/test/java/org/wso2/carbon/identity/user/endpoint/impl/MeApiServiceImplTest.java]: As this is a test method do we really need to add method comments. If so we need to add method comments to all the test methods as we have not added them :)

File: components/org.wso2.carbon.identity.user.endpoint/src/test/java/org/wso2/carbon/identity/user/endpoint/impl/ValidateCodeApiServiceImplTest.java (hunk 6 above)
Comment (DMHP) [reply components/org.wso2.carbon.identity.user.endpoint/src/test/java/org/wso2/carbon/identity/user/endpoint/impl/ValidateCodeApiServiceImplTest.java]: Actually from ValidateCodeApiServiceImpl.java we have handled this exception by using a try catch block. So we this exception is not throwing to this level. So we don't need to expect the exception here

**Pull request merged**